src/txrequest.h

```diff
@@ -0,0 +1,347 @@
+
+    //! Wrapper around Index::...::erase that keeps m_peerinfo up to date.
+    template<typename Tag>
+    typename Index::index<Tag>::type::iterator Erase(typename Index::index<Tag>::type::iterator it)
+    {
+        auto peerit = m_peerinfo.find(it->m_peer);
+        peerit->second.m_requested -= it->GetState() == State::REQUESTED;
+        if (--peerit->second.m_total == 0) m_peerinfo.erase(peerit);
+        return m_index.get<Tag>().erase(it);
+    }
+
+    //! Wrapper around Index::...::modify that keeps m_peerinfo up to date.
+    template<typename Tag, typename Modifier>
+    void Modify(typename Index::index<Tag>::type::iterator it, Modifier modifier)
+    {
+        auto peerit = m_peerinfo.find(it->m_peer);
+        peerit->second.m_requested -= it->GetState() == State::REQUESTED;
+        m_index.get<Tag>().modify(it, std::move(modifier));
+        peerit->second.m_requested += it->GetState() == State::REQUESTED;
+    }
+
+    //! Change the state of an entry to something non-IsSelected(). If it was IsSelected(), the next best entry will
+    //! be marked CANDIDATE_BEST.
+    void ChangeAndReselect(typename Index::index<ByTxid>::type::iterator it, State new_state);
+
+    //! Convert a CANDIDATE_DELAYED entry into a CANDIDATE_READY. If this makes it the new best CANDIDATE_READY (and no
+    //! REQUESTED exists) and better than the CANDIDATE_BEST (if any), it becomes the new CANDIDATE_BEST.
+    void PromoteCandidateNew(typename Index::index<ByTxid>::type::iterator it);
+
+    //! Convert any entry to a COMPLETED one. If there are no non-COMPLETED entries left for this txid, they are all
+    //! deleted. If this was a REQUESTED entry, and there are other CANDIDATEs left, the best one is made
+    //! CANDIDATE_BEST. Returns whether the Entry still exists.
+    bool MakeCompleted(typename Index::index<ByTxid>::type::iterator it);
+
+    //! Make the data structure consistent with a given point in time:
+    //! - REQUESTED entries with exptime <= now are turned into COMPLETED.
+    //! - CANDIDATE_DELAYED entries with reqtime <= now are turned into CANDIDATE_{READY,BEST}.
+    //! - CANDIDATE_{READY,BEST} entries with reqtime > now are turned into CANDIDATE_DELAYED.
+    void SetTimePoint(std::chrono::microseconds now);
+
+public:
+    //! Construct a TxRequestTracker with the specified parameters.
+    TxRequestTracker(std::chrono::microseconds timeout, bool deterministic = false);
+
+    // Disable copying and assigning (a default copy won't work due the stateful EntryTxidExtractor).
+    TxRequestTracker(const TxRequestTracker&) = delete;
+    TxRequestTracker& operator=(const TxRequestTracker&) = delete;
+
+    //! A peer went offline, delete any data related to it.
+    void DeletedPeer(uint64_t uint64_t);
+
+    //! For whatever reason, we no longer need this txid. Delete any data related to it.
+    void AlreadyHaveTx(const uint256& txid);
+
+    //! We received a new inv, enter it into the data structure.
```

Comment:
259fb92

We call `ReceivedInv` on the behalf of a peer sending us an orphan tx. Is following scenario plausible ?
* Malicia sends N txn to Alice and N' conflicting to Bob
* Malicia sends txB child spending from all parent of N to Alice
* Alice sends txB to Bob, at reception Bob calls `RequestTx/ReceivedInv` on behalf of Alice
* Malicia repeats this trick until reaching Alice's  MAX_PEER_TX_ANNOUNCEMENTS (and also avoid hitting MAX_STANDARD_TX_WEIGHT for orphan)
* When Bob calls `GetRequestable` at GETDATA sending to Alice, only MAX_GETDATA_SZ will be effectively processed, the rest obstructing Alice's announced invs until GETDATA_TX_INTERVAL expiration
* Alice sends transaction for the MAX_GETDATA_SZ set
* Malicia may repeat timely repeat trick to maintain transaction propagation blocking anticipating Alice-Bob round-trips

Attack might seem costly because you have to fulfill Alice mempool with parent transaction but making them conflicted with the rest of network mempools avoid to effectively pay their feerate.

Reply:
I think you're confused. `MAX_GETDATA_SZ` doesn't stop processing; it only makes the request be split up into multiple GETDATA messages.

Reply:
You're right on `MAX_GETDATA_SZ`.

Still, do you think that a third-party can delay transaction announcement between Alice and Bob by forcing her to send him orphans until reaching MAX_PEER_TX_ANNOUNCEMENTS ?

Reply:
I need to think about this more.